### ISSUE TITLE
Interrupting stuck network processes

### ISSUE THREAD

**Issue by Mellvik** [ghaerr/elks#1356]
My intention was to put this on the 0.7.0 list, but it may as well be considered a regular bug.

My memory tells me we've touched this one before, but Seraching the repository didn't give me any hints.

Anyway, stuck network processes hardly ever react to signals - kill -9 ... whatever. Say we do a 'net stop' while an incoming telnet is active and the telnetd process is stuck forever, and reboot is required. 

There are many such examples- in particular when running elks-to-elks networking because of timeouts that eventually ends up in hung processes, but this is the easiest to reproduce. 

Attacking this problem would be really helpful. And I'm interested in doing so. @ghaerr, any ideas about where to start?

-M

**Comment by ghaerr** [comment]
This is going to be a really tough one - the biggest reason is that even if we "fix" the `interruptible_sleep()` that the processes are likely waiting on so that they "return out of the kernel" (and are killed), the kernel network routines will be left with their semaphores in possibly incorrect/bad states (e.g. a semaphore may be left in the always off or on position, which will fail when networking is "restarted".)

In order to see this, I advise for you to start by looking at elks/net/ipv4/af_inet.c: each of the "interruptible_sleep" routines will need to be unwound properly, on a case-by-case basis. Remember how sleep works, a process is effectively unscheduled until a wake_up call is made, except, if the process receives a signal, the process will wake, and its kernel task will continue; this means it will return from interruptible_sleep, and code can check current->signal to see whether the wakeup was a result of a wake_up call, or a signal. THEN, each individually-coded sleep can be possibly unwound properly.

Lets take inet_bind() in the above file to start: the `rwlock` semaphore is DOWN, so that would have to be reversed, and ktcp has been given a request, for which there has (presumably) not been a response. We also need to cancel `tcpdev_clear_data_avail()`. After all this, we could then allow the process to be killed, which occurs after the current system call is completed, but if ktcp is still running, it will be put into a bad state, so it needs to be killed also, and it may be sleeping in a similar kernel routine, with the same or different set of semaphores gated/unlocked/etc.

This whole dilemma is the reason many UNIXs and sometimes Linux still hang and require a reboot, despite repeated kill requests. In some sense, what is needed is a kind of "kernel reset" that resets all network variables to their starting state, and kills all associated process - a big kluge, and not always possible since the kernel doesn't "know" which processes might be "network" processes.

Take a look and we'll go from here after you've looked further at it. Frankly, even if we "fix" it, we really can't guarantee kernel correctness afterwards, which might cause very strange problems for ourselves and users after a kill.

**Comment by Mellvik** [comment]
Thank you @ghaerr - 
that was a cold shower indeed. Very thorough - much appreciated.

Maybe we could lower the level of ambition, ignore the havoc a -9 signal may cause, and - say -enable ktcp to terminate all open connections when terminating, regardless of cause.
My case is mostly ´net stop´, forgetting that there are open connections, idle or active.

I'll take a look at the code you suggested in an attempt to understand the big picture. _.)

-M

**Comment by ghaerr** [comment]
> enable ktcp to terminate all open connections when terminating, regardless of cause.

Well, all sockets opened are automatically closed by ktcp when it exits. I don't think this currently changes any processes sleeping on their own socket connections to the kernel though. However, knowing that there is no /dev/tcpdev open (meaning ktcp has exited) may allow for a killed process to exit its kernel task state more cleanly, without worry for ktcp corruption as mentioned above.

That would mean one would execute `net stop`, then possibly kill the remaining network processes... is that what you're doing now, when seeing the hanging processes?

**Comment by Mellvik** [comment]
> enable ktcp to terminate all open connections when terminating, regardless of cause.
> 
> Well, all sockets opened are automatically closed by ktcp when it exits. I don't think this currently changes any processes sleeping on their own socket connections to the kernel though. However, knowing that there is no /dev/tcpdev open (meaning ktcp has exited) may allow for a killed process to exit its kernel task state more cleanly, without worry for ktcp corruption as mentioned above.
> 
> That would mean one would execute net stop, then possibly kill the remaining network processes... is that what you're doing now, when seeing the hanging processes?
> 
Yes, that's it. And yes, it does seem like a fairly common situation - the device going away. Like the carrier going away on a serial/modem except in this case there may be many processes using it...
Anyway, getting this one to work would be a big step.

-M

**Comment by cml37** [comment]
Would implementing support for socket send and receive timeouts help here?  I've noticed, for example, with the httpd server, if a client socket doesn't terminate gracefully, the socket will forever be stuck.

**Comment by ghaerr** [comment]
Hi @cml37,

We already implement send and receive timeouts in `ktcp`, the user-mode application that actually implements the TCP/IP protocols. Since ELKS applications have to communicate with the kernel through (socket) syscalls, the kernel acts as a go-between between network applications and ktcp itself. The problem we're seeing above is that in certain circumstances, a network application will get permanently put to sleep by the kernel and for various reasons, it is very hard to unwind it without killing the process and/or getting out of sync with ktcp. More work needs to be done on this.

> I've noticed, for example, with the httpd server, if a client socket doesn't terminate gracefully, the socket will forever be stuck.

Is this in regards to the very cool video Sean at Action Retro recently produced (linked in https://github.com/ghaerr/elks/discussions/2606)? Sean seems to have run into some memory-limitation issues of ktcp while offering his ELKS httpd server to the world, where it quickly runs out of memory. We can probably increase the amount of memory available to ktcp for some issues seen in Sean's video, but memory is still very limited, as ktcp currently only has a 64K max data segment for which to hold all TCP connection data. 

I'd like more information on what you're seeing regarding non-graceful socket terminations and how details on whether ktcp or httpd is stuck, or whether its just a IP/port number that is permanently allocated, etc.

Thank you!

**Comment by cml37** [comment]
> Is this in regards to the very cool video Sean at Action Retro recently produced (linked in https://github.com/ghaerr/elks/discussions/2606)? Sean seems to have run into some memory-limitation issues of ktcp while offering his ELKS httpd server to the world, where it quickly runs out of memory.

It sure is!  I decided to look at this a little bit, and noticed that what happens is the forked process is getting "stuck".  I actually found that killing the forked process would unjam the application as well.  So, I made a special build of httpd to not fork at all to see what happens, and thus the whole httpd application would get stuck whenever we ended up with non graceful termination, or so it appears.  See [here](https://x.com/RetroTechChris/status/2018079177233063989).  I was naturally incorrect about the socket timeouts piece if that is being supported in ktcp.

**Comment by ghaerr** [comment]
@cml37 
Nice idea playing around with a non-fork rewrite of httpd! I'm not quite sure yet what the problem is exactly, and plan on setting up a test scenario to figure it out. I must say we've been working mostly on just getting things to work, while apparently not looking closely enough at performance, but it is very nice to see ELKS used as was seen in Sean's video - and was glad to see it actually worked for a while, until resources were used up.

I am guessing what is actually happening is that `httpd` has a listen socket open, which naturally will then at the TCP/IP stack level in the user mode `ktcp` application attempt to accept any and all incoming requests (without limit) and allocate space for the connection, while then notifying the upper level `httpd` of each request by returning from the blocked `accept` call. Ktcp is then running out of memory and may not be properly resetting the remote connection when out of memory - I will have to check. However, there currently is no limit to the number of outstanding connection requests, all are immediately accepted, regardless of how fast `httpd` can process accepts.

There are multiple potential problems here: ktcp's (fixed) heap is currently set to 33k bytes, it can be increased using `chmem` to almost 64k, which will increase the number of open connections possible. It's not shipped with 64K heap in order to save 32K RAM when not normally needed. But ktcp doesn't remember when it's out of memory and continues to try to accept any further connection requests - that needs fixing. And the bigger problem is that httpd (and ELKS itself) is limited to a max 16 processes total running at any time. If/when a fork fails, the just-accepted connection may not be getting taken down properly, although I see the socket is closed by httpd.

Lots to look into. I plan on looking further into this and would like to be able to test while running on QEMU. A test harness needs to be setup for sending connection requests quickly in order to more accurately pinpoint where things first start to go wrong.

**Comment by djohanse** [comment]
I'm not elks-network-savvy enough to work code changes for this, but as a reference point (and for something to do on a Saturday morning) I fed Melvikks initial question and ghaerr's first response into openCode based on todays (8/9/2026) master.

Here's what open code thinks:

Analysis complete. Here's what I found after tracing the full syscall → tcpdev → ktcp path.
Why telnetd hangs forever (and why even kill -9 fails)
The root mechanism is confirmed. A process only gets a pending signal acted upon inside do_signal(), which is called from exactly two places: after a syscall returns (irqtab.S:186) or when an interrupt hits a user-mode task (irqtab.S:315). A task sleeping inside a syscall never reaches either — so it never exits, and its signal (even SIGKILL) sits pending forever.
For the net stop reproduction:
1. net stop (elkscmd/rootfs_template/bin/net:66) SIGTERMs the pids matching ktcp|telnet|httpd|ftpd — killing ktcp and the telnetd daemon/child.
2. The telnetd child is in client_loop(), typically blocked in a socket write() (login banner, shell prompt, echo) → inet_write() → af_inet.c:359-362:
while (bufin_sem == 0) {
    interruptible_sleep_on(sock->wait);
}
This loop never checks current->signal. ktcp is dead, so bufin_sem stays 0 forever. Each signal just wakes the task, the loop re-arms, and the syscall never returns to user space.
3. Worse, the scheduler turns this into a busy-spin: sched.c:93-101 keeps any TASK_INTERRUPTIBLE task with a pending signal on the runqueue. Once signaled, __sleep_on returns immediately on every pass (schedule() refuses to put it to sleep), so the task burns CPU in a tight reschedule loop instead of sleeping.
4. The child holds the global rwlock (af_inet.c:345) across that sleep. It's now leaked at -1. Every subsequent down(&rwlock) from any other process goes into sleep_on() (sleepwake.c:163, uninterruptible) — those processes are then also unkillable.
Result: exactly what the message describes — one stuck telnetd, and networking dead until reboot even across net stop/net start (the leaked rwlock survives ktcp's reset because tcpdev_release() only resets bufin_sem/bufout_sem, tcpdev.c:158-164).

<img width="1086" height="298" alt="Image" src="https://github.com/user-attachments/assets/d7dec4b2-7289-4f8f-bf42-6b6ae4939a21" />


tcpdev.c:50 (ktcp's read) does check the signal (:51), and af_unix.c/pipe.c/socket.c:241 all check signals properly — so this is specifically the INET/tcpdev path. The three missing signal checks (bind, listen, read-2nd-loop, write) plus the retry loop are the kill sites.
The state-corruption inventory (your semaphore concern, confirmed)
Even after adding signal checks, each unwind must repair, and today the protocol makes that hard:
- rwlock — leaked by bind/listen/read/write if killed mid-sleep. Needs up() on unwind.
- sock->sem — taken at af_inet.c:307; released :320. Only held briefly (no sleep between), so lower risk.
- bufin_sem / bufout_sem — the tcpdev buffers are single-buffered and global; only one request may be in flight. The "response" for a dead request will still arrive later (tcpdev_write → inet_process_tcpdev → wake_up(sock->wait)), and nothing will consume it. The next request would then read a stale tdin_buf meant for the dead socket — the TDT_RETURN/TDT_BIND/etc. payload carries sock, so a response matcher is possible in principle, but the kernel currently has no notion of "outstanding request" to match against.
- ktcp's TCB — the request side (TDC_BIND/TDC_CONNECT/TDC_ACCEPT, e.g. tcpdev_bind creating a TCB and not responding until a SYN arrives for accept) leaves orphaned/半-bound TCBs in ktcp keyed by a dead sock pointer.
- The reverse direction: if the kernel-side process ever did die mid-transaction (after the fix), ktcp itself would block forever in kernel tcpdev_write → down(&bufin_sem) (tcpdev.c:102, uninterruptible) waiting for the dead process to call tcpdev_clear_data_avail(). Then ktcp is hung in-kernel and also unkillable — the exact "kill ktcp and it's in a bad state" problem.
Elks-to-elks "timeout" hang
Also confirmed. When a peer stops ACKing, ktcp's retransmission memory grows until TCP_RETRANS_MAXMEM (5120, tcp.h:57), then tcpdev_write returns -ERESTARTSYS (elkscmd/ktcp/tcpdev.c:380-384, :416-427). inet_write responds by sleeping 100 ms and retrying inside the syscall (af_inet.c:372-378) with no signal check — an infinite, unkillable retry loop. This is a second, distinct path to the same "stuck forever" state.
Where I'd start (analysis, not yet code)
Your proposed direction is right, but the audit suggests a two-layer fix rather than purely case-by-case:
1. Make every sleep killable — add the missing current->signal checks to the four sites above, and to the ERESTARTSYS retry loop. On signal, unwind: up(&rwlock) where held, return a restart-able error. This alone fixes the "unkillable process" half.
2. Solve the stale-response/half-transaction problem structurally — this is the part that "can't be guaranteed correct" case-by-case. The clean primitive is a per-socket request-in-flight marker (the protocol is already synchronous — one outstanding request per socket, serialized by rwlock), so that when a transaction is aborted, the kernel can recognize and discard the late response (inet_process_tcpdev already receives the sock). Combined with: on tcpdev_release (ktcp death), the kernel should abort all in-flight socket operations — wake every socket waiter, force the bufin_sem/rwlock loops to see a "network down" condition and unwind — which is the "kernel reset" kludge, scoped to INET sockets where the kernel does actually know the sockets (pops/socket table).
The honest conclusion of the analysis matches the message: after any kill, kernel correctness for the tcpdev protocol can't be guaranteed without a coordinated reset, because the single-buffered shared-memory protocol (one tdout_buf, one tdin_buf, two global semaphores) has no recovery mechanism for lost transactions. The realistic target is "killable processes + no leaked locks + dropped stale responses", achieved by (1) and a narrow version of (2) — which is likely what makes net stop reproducible-recoverable without a reboot.

**Comment by ghaerr** [comment]
Thanks @djohanse. The AI's long response confirms my first response that unfortunately, a majority of the tcpdev shim code in between the kernel and ktcp would have to be rewritten in order to properly unwind the multiple semaphores properly. As AI also confirmed, this is extremely complicated given the tenuous primarily serialized/synchronous relationship between ktcp and tcpdev as well. Because of this, even writing extra code to unblock "locked" processes still won't allow ktcp to recover, and the tcpdev driver has to be properly closed to reopen without error. Long list of issues.

> Worse, the scheduler turns this into a busy-spin:

I was not aware of this potential scheduling spin-loop and will look into it. I have seen rare cases in the past where the system slows to a crawl, with applications displaying console messages very slowly. I'm wondering if this or a similar issue might be the reason for that, with ELKS spending lots of time "running" a spin-locked process. If so, this could result in a major bug fix (or at least real understanding given the fact the process involved can't actually be easily killed!!).

**Comment by ghaerr** [comment]
After reading this more carefully, there may be a way to allow a `net stop` to work as intended "most-of-the-time". That is, fixing the "normal" cases of processes sleeping on a socket read or write would allow them to process any sent signal, after first repairing the normal rwlock semaphore from being killed by a `kill` in `net stop`. Both kernel networking and ktcp could likely be in an incorrect state, but this could possibly be salvaged by having `net stop` wait a second to kill `ktcp` until after all network child processes have been killed first. The tcpdev driver close routine, executed after ktcp exits, would then reset all network semaphores so that a subsequent `net start` would allow everything to be restarted. It is not clear whether ktcp could be correctly terminated without trashing the kernel, this would have to be tested.

I'll have to think about this a bit: as AI mentions, allowing any process to be unstuck puts the kernel at risk for incorrectness, and allowing such could allow the kernel to continue with major issues unbeknownst to the user. This would have to be weighed against the utility of  allowing a typical `net stop` (e.g. no actually active network activity) to work in most circumstances. A big problem will be that users don't know about any of this so they'll issue a `net stop` whenever they like, the system becomes unstable, and the user then complains the system crashed, rather than just having stuck processes and rebooting.

Comments?

**Comment by djohanse** [comment]
Fed your comments to AI.  His thoughts:

Your plan is sound, and the ordering (children first, then ktcp) is actually required, not just helpful. The reason is subtle: tcpdev_release resets bufin_sem to 0, but a reset to 0 doesn't wake anyone — processes stuck in while (bufin_sem == 0) (af_inet.c:301, :359) stay asleep forever. So ktcp can only die safely after the children are gone.

On your specific points:

"Can ktcp be terminated without trashing the kernel?" — Killing ktcp is not a kernel-corruption risk; it's an ordinary user-space process and dies at the next syscall/interrupt boundary. The only genuine hazard is protocol state, and there is exactly one real hang: ktcp's own down(&bufin_sem) at tcpdev.c:102 is uninterruptible (sleepwake.c:163). If a child died mid-response (leaving bufin_sem == -1, since the TDT_RETURN/TDT_ACCEPT/TDT_BIND paths rely on the woken process to call tcpdev_clear_data_avail, af_inet.c:85), then ktcp's next write blocks forever and even kill can't unstick it. That down must become signal-aware for the plan to be airtight.

rwlock repair — With Part 1, the rwlock self-heals: the abort path does up(&rwlock) plus a conditional up(&bufout_sem) (only when < 0, meaning ktcp hasn't consumed the command). A rwlock = 0 reset in tcpdev_release is cheap insurance but mostly redundant once unwinding exists. Note one more leak your Part 1 must cover: inet_accept returns -ERESTARTSYS on signal at af_inet.c:246 without releasing bufout_sem — a retry would deadlock in get_tdout_buf().

The correctness-risk concern — I'd frame it this way:
- Today the kernel is permanently corrupted on net stop (rwlock stuck at −1, reboot required). Part 1 + reset makes the window bounded and recoverable, so it strictly reduces incorrectness risk.
- The real silent-danger is the stale-response case: a killed mid-read process leaves bufin_sem = -1 with an orphaned response in tdin_buf, and the next socket op consumes the wrong response and then hangs waiting for its own. Your ordering contains this — but only if net is restarted. That's the "user kills one process and keeps going" case Part 2 (per-socket request tracking, plus checking !tcpdev_inuse in the wait loops → -ENETDOWN) would close.
- On UX: a broken net stop is guaranteed confusion. Making it work "most of the time" while treating arbitrary mid-activity kills as unsupported is strictly better. net stop/restart becomes the supported lifecycle; the residual edge case should just be documented.

Recommended concrete path:
1. af_inet.c: signal checks + lock unwind in read :301, write :359, the ERESTARTSYS retry :372, and accept :244.
2. tcpdev.c:102: make the down(&bufin_sem) interruptible so ktcp is always killable.
3. tcpdev_release: also reset rwlock (belt-and-suspenders).
4. bin/net: split the kill — children first, sleep 1, then ktcp (rootfs_template/bin/net:66 currently kills both in one line).
5. Test idle net stop/restart first; then a mid-transfer kill to characterize the residual behavior before deciding on Part 2.

**Comment by ghaerr** [comment]
Well, I'm glad the AI tool confirms my observations, except for one point about "strictly better":

>  broken net stop is guaranteed confusion. Making it work "most of the time" while treating arbitrary mid-activity kills as unsupported is strictly better. net stop/restart becomes the supported lifecycle; the residual edge case should just be documented.

Users "confused" that `net stop` didn't do anything isn't always worse than documenting the "residual edge case" which becomes the main (unknown since user don't read docs) case every time a user tries stopping the network when networking is active, then finding the system hung, rather than just stuck processes. 

That said, there's currently quite a few other cases where networking gets messed up enough to require a reboot: see #2801 (I'm preparing separate network bug and issues list which will require fixing each one individually along with regression testing to ensure each fix doesn't break something else). 

So - it does probably make some sense to continue thinking about this, since the complete inability to interrupt stuck processes without a reboot is also a major problem. Another option I'm considering would be a way to instruct the kernel or ktcp to stop processing network requests, so that ktcp could be stopped with network activity at a known standstill.

**Comment by Mellvik** [comment]
Thanks, @djohanse - I find this really interesting and helpful - and will be using it when taking a(nother) deep dive into `ktcp` after wrapping up the current project on TLVC.

Completely off topic, but here we go: @ghaerr, there seem to be a serious problem with HD minix filesystem generation during system build (both ELKS and TLVC): 32M and 64M images get created with 2048 nodes. What's even worse is that if the inode table fills up (2047 used), the filesystem is damaged beyond repair.  I'm working on a fix and figure you're likely to be more versed in the filesystem code than I am.

**Comment by ghaerr** [comment]
> Completely off topic, but here we go

Please open a new issue as this one is already quite long. More details on what is seen when "damaged upon repair" as well as whether the inode depletion can be duplicated on floppy would help.
